src/rdma.c

```diff
@@ -0,0 +1,1894 @@
+
+    /* Force binding of 0.0.0.0 if no bind address is specified. */
+    if (listener->bindaddr_count == 0) {
+        bindaddr_count = 2;
+        bindaddr = default_bindaddr;
+    }
+
+    listener->priv = rdma_listener = zcalloc_num(bindaddr_count, sizeof(*rdma_listener));
+    for (j = 0; j < bindaddr_count; j++) {
+        char *addr = bindaddr[j];
+        int optional = *addr == '-';
+
+        if (optional) addr++;
+        if (strchr(addr, ':')) {
+            /* Bind IPv6 address. */
+            ret = rdmaServer(server.neterr, port, addr, AF_INET6, rdma_listener);
+        } else {
+            /* Bind IPv4 address. */
+            ret = rdmaServer(server.neterr, port, addr, AF_INET, rdma_listener);
+        }
+
+        if (ret == ANET_ERR) {
+            serverLog(LL_WARNING, "RDMA: Could not create server for %s:%d: %s", addr, port, server.neterr);
+
+            return C_ERR;
+        }
+
+        int fd = rdma_listener->cm_channel->fd;
+        anetNonBlock(NULL, fd);
+        anetCloexec(fd);
+        listener->fd[listener->count++] = fd;
+        rdma_listener++;
+    }
+
+    return C_OK;
+}
+
+static int connRdmaAddr(connection *conn, char *ip, size_t ip_len, int *port, int remote) {
+    rdma_connection *rdma_conn = (rdma_connection *)conn;
+    struct rdma_cm_id *cm_id = rdma_conn->cm_id;
+    struct sockaddr_storage *ss = NULL;
+    struct sockaddr_in *sa4;
+    struct sockaddr_in6 *sa6;
+
+    if (remote)
+        ss = (struct sockaddr_storage *)rdma_get_peer_addr(cm_id);
+    else
+        ss = (struct sockaddr_storage *)rdma_get_local_addr(cm_id);
+
+    if (!ss) {
+        goto error;
+    }
+
+    if (ss->ss_family == AF_INET) {
+        sa4 = (struct sockaddr_in *)ss;
+        if (ip) {
+            if (inet_ntop(AF_INET, (void *)&(sa4->sin_addr), ip, ip_len) == NULL) {
+                goto error;
+            }
+        }
+
+        if (port) {
+            *port = ntohs(sa4->sin_port);
+        }
+    } else if (ss->ss_family == AF_INET6) {
+        sa6 = (struct sockaddr_in6 *)ss;
+        if (ip) {
+            if (inet_ntop(AF_INET6, (void *)&(sa6->sin6_addr), ip, ip_len) == NULL) {
+                goto error;
+            }
+        }
+
+        if (port) {
+            *port = ntohs(sa6->sin6_port);
+        }
+    } else {
+        /* TODO IB protocol */
+        goto error;
+    }
+
+    return 0;
+
+error:
+    if (ip) {
+        if (ip_len >= 2) {
+            ip[0] = '?';
+            ip[1] = '\0';
+        } else if (ip_len == 1) {
+            ip[0] = '\0';
+        }
+    }
+
+    if (port) *port = 0;
+
+    return -1;
+}
+
+static void rdmaInit(void) {
+    pending_list = listCreate();
+
+    VALKEY_BUILD_BUG_ON(sizeof(ValkeyRdmaFeature) != 32);
+    VALKEY_BUILD_BUG_ON(sizeof(ValkeyRdmaKeepalive) != 32);
+    VALKEY_BUILD_BUG_ON(sizeof(ValkeyRdmaMemory) != 32);
+    VALKEY_BUILD_BUG_ON(sizeof(ValkeyRdmaCmd) != 32);
+
+    if (ibv_fork_init()) {
```

Comment:
I am concerned about using RDMA when the Valkey system is configured in a primary+replica configuration.  My understanding is that ibv_fork_init only makes RDMA fork-friendly, not truly fork-safe.

When a replica connects to the primary, the primary forks and its child generates a full RDB dump and sends it to the replica.  This takes a significant amount of time and relies on copy-on-write semantics so that the child of the primary can send a consistent snapshot of the primary's state at the time of the fork to the replica.

Has this scenario been tested?

Note: https://www.rdmamojo.com/2012/05/24/ibv_fork_init/ says "The fork() support in libibverbs isn't full and it assumes that only the parent process will do the RDMA operations. If one of the child processes will try to perform RDMA operations, it may experience various problems."  This appears to correspond quite well to the primary-replica scenario.

Reply:
> I am concerned about using RDMA when the Valkey system is configured in a primary+replica configuration. My understanding is that ibv_fork_init only makes RDMA fork-friendly, not truly fork-safe.
> 
> When a replica connects to the primary, the primary forks and its child generates a full RDB dump and sends it to the replica. This takes a significant amount of time and relies on copy-on-write semantics so that the child of the primary can send a consistent snapshot of the primary's state at the time of the fork to the replica.
> 
> Has this scenario been tested?
> 

A valkey-server is connected by clients via RDMA, several cases have been tested:
* run `bgsave` command in a client tool
* auto save after lots of `set` commands
* another valkey-server runs `REPLICAOF` command via TCP connection.

All these case work fine.

I guess after RDMA gets supported, `REPLICAOF` command may support a new parameter `transport` to enable RDMA. Then we can use a RDMA connection to exchange data between primary and replica.

> Note: https://www.rdmamojo.com/2012/05/24/ibv_fork_init/ says "The fork() support in libibverbs isn't full and it assumes that only the parent process will do the RDMA operations. If one of the child processes will try to perform RDMA operations, it may experience various problems." This appears to correspond quite well to the primary-replica scenario.

This link has already explained `ibv_fork_init` and `fork syscall`, luckly the bgsave child process will not read/write via connections. (In the current implementation, once the child process tries to perform RDMA operations, it will hit a segment fault)

I have indeed hit a curious issue during developing this work. Please see [code](https://github.com/pizhenwei/valkey/blob/feature-rdma/src/rdma.c#L180C14-L180C33):
```
/* To make Valkey forkable, buffer which is registered as RDMA
 * memory region should be aligned to page size. And the length
 * also need be aligned to page size.
 * Random segment-fault case like this:
 * 0x7f2764ac5000      -      0x7f2764ac7000
 * |ptr0 128| ... |ptr1 4096| ... |ptr2 512|
 *
 * After ibv_reg_mr(pd, ptr1, 4096, access), the full range of 8K
 * becomes DONTFORK. And the child process will hit a segment fault
 * during access ptr0/ptr2.
 * Note that the memory can be freed by libc free only.
 * TODO: move it to zmalloc.c if necessary
 */
static void *page_aligned_zalloc(size_t size) {
    void *tmp;
    size_t aligned_size, page_size = sysconf(_SC_PAGESIZE);

    aligned_size = (size + page_size - 1) & (~(page_size - 1));
    if (posix_memalign(&tmp, page_size, aligned_size)) {
        serverPanic("posix_memalign failed");
    }

    memset(tmp, 0x00, aligned_size);

    return tmp;
}
```

Once unaligned memory area is used for RDMA MR, the child process will crash randomly. We need make sure all the memory areas used by RDMA MR must be page size aligned.

After this issue got fixed , I didn't see any other issues during test. And I also dived into the linux kernel inifiniband core driver and rdma-core, I didn't find any obvious proplems on RDMA memory management.
Do you have any known issues? Please let me know, I'll fix it ASAP.

Reply:
> A valkey-server is connected by clients via RDMA, several cases have been tested:
>     * run `bgsave` command in a client tool
>     * auto save after lots of `set` commands
>     * another valkey-server runs `REPLICAOF` command via TCP connection.

Only the last one comes close to the scenario I described.  However, in the last one it appears that the REPLICA was still receiving the RDB-dump via TCP.  Could you point out where this special handling is done?

> > Note: https://www.rdmamojo.com/2012/05/24/ibv_fork_init/ says "The fork() support in libibverbs isn't full and it assumes that only the parent process will do the RDMA operations. If one of the child processes will try to perform RDMA operations, it may experience various problems." This appears to correspond quite well to the primary-replica scenario.
> 
> This link has already explained `ibv_fork_init` and `fork syscall`, luckly the bgsave child process will not read/write via connections.

The BGSAVE command causes the child to simply write to disk, so there is no problem.  However, the synchronization between a replica and its primary causes the primary's BGSAVE-child to write to a socket connection to feed the RDB-dump to the replica before using the same connection to feed the replication stream to the replica.

> (In the current implementation, once the child process tries to perform RDMA operations, it will hit a segment fault)

If I understand you correctly, this is exactly the problem that lead to this conversation.

Reply:
> > A valkey-server is connected by clients via RDMA, several cases have been tested:
> > * run `bgsave` command in a client tool
> > * auto save after lots of `set` commands
> > * another valkey-server runs `REPLICAOF` command via TCP connection.
> 
> Only the last one comes close to the scenario I described. However, in the last one it appears that the REPLICA was still receiving the RDB-dump via TCP. Could you point out where this special handling is done?
> 
> > > Note: https://www.rdmamojo.com/2012/05/24/ibv_fork_init/ says "The fork() support in libibverbs isn't full and it assumes that only the parent process will do the RDMA operations. If one of the child processes will try to perform RDMA operations, it may experience various problems." This appears to correspond quite well to the primary-replica scenario.
> > 
> > 
> > This link has already explained `ibv_fork_init` and `fork syscall`, luckly the bgsave child process will not read/write via connections.
> 
> The BGSAVE command causes the child to simply write to disk, so there is no problem. However, the synchronization between a replica and its primary causes the primary's BGSAVE-child to write to a socket connection to feed the RDB-dump to the replica before using the same connection to feed the replication stream to the replica.

The child process can't use existed connection currently. This is the limitation now. Luckly, Valkey can listen on both TCP and RDMA. TCP is optional for replica connection.

Once this feature is merged, I think I can post kernel patch and rdma-core patch to resolve it. The basic idea is: introduce a new syscall(like `madvise(void addr, size_t length, MADV_CHILDFORK)`), then the RDMA MR would be used in the child only, the parent process would release this MR after fork.

Reply:
I don't think the child process would need to use RDMA. Even in the context of #60, I would assume the new connection between the child and the replica  would be over TCP. So is it correct understanding that if the child process never uses RDMA, we don't need more kernel/rdma patches?

Also is the minimum kernel version and hardware requirement documented somewhere?